src/networking.c

```diff
@@ -3246,7 +3246,7 @@ char *getClientSockname(client *c) {
 int isClientConnIpV6(client *c) {
     /* The cached client peer id is on the form "[IPv6]:port" for IPv6
      * addresses, so we just check for '[' here. */
-    if (c->conn->type == NULL && server.current_client) {
+    if ((c->conn == NULL || c->conn->type == NULL) && server.current_client) {
```

Comment:
couple of thoughts about this function:

1. This is still not very safe because it assumes `server.current_client` is set and is a real client (btw aof has already violated this in `loadSingleAppendOnlyFile`).
3. I am not sure about the fall back to `current_client` (assuming it is set and real). The caller is asking for `c`'s IPv6 capability not for `current_client`. This implicit conversion can cause surprises in the future IMO.

Overall I feel that this logic is brittle. On a high level, I think we should consider formalizing the concept of `fake` clients, as opposed to implying it from `conn->type` or something else. Then this logic will be as straightforward as ` return !isFakeClient(c) && getClientPeerId(c)[0] == '['`

Reply:
you are right, this logic is indeed fragile. The fake client i can think of:
1. aof fake client, the id is CLIENT_ID_AOF
2. lua/function fake client, has a script client flag.
3. module fake client, has a module client flag (or no flag).
4. a fake (or cached) client we used in CLUSTER SLOTS (has the conn and the conn-type is NULL)

we do need a isFakeClient function i think.

Reply:
A new fake_client client flag or a new isFakeClient function? `(c->id == CLIENT_ID_AOF || c->flag.script || c->flag.module || (c->conn && c->conn->type == NULL));`

Reply:
I would prefer a new flag. We could even save a new function. This list is quite hard to reason about IMO.